libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -641,176 +684,441 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
 template <typename _Ty>
 _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty&) {return true;}
 
-template<_CUDA_VSTD::size_t _Alignment>
+// __is_local_smem_barrier returns true if barrier is (1) block-scoped and (2) located in shared memory.
+template<thread_scope _Sco, typename _CompF, bool _Is_mbarrier = (_Sco == thread_scope_block) && _CUDA_VSTD::is_same<_CompF, _CUDA_VSTD::__empty_completion>::value>
 _LIBCUDACXX_INLINE_VISIBILITY
-inline void __strided_memcpy(char * __destination, char const * __source, _CUDA_VSTD::size_t __total_size, _CUDA_VSTD::size_t __rank, _CUDA_VSTD::size_t __stride = 1) {
-    if (__stride == 1) {
-        memcpy(__destination, __source, __total_size);
-    }
-    else {
-        for (_CUDA_VSTD::size_t __offset = __rank * _Alignment; __offset < __total_size; __offset += __stride * _Alignment) {
-            memcpy(__destination + __offset, __source + __offset, _Alignment);
-        }
-    }
+bool __is_local_smem_barrier(barrier<_Sco, _CompF> & __barrier) {
+    NV_IF_ELSE_TARGET(
+        NV_IS_DEVICE, (
+            return _Is_mbarrier && __isShared(&__barrier);
+        ),(
+            return false;
+        )
+    );
 }
 
-template<_CUDA_VSTD::size_t _Alignment, bool _Large = (_Alignment > 16)>
-struct __memcpy_async_impl {
-    _LIBCUDACXX_DEVICE static inline async_contract_fulfillment __copy(char * __destination, char const * __source, _CUDA_VSTD::size_t __total_size, _CUDA_VSTD::size_t __rank, _CUDA_VSTD::size_t __stride) {
-        __strided_memcpy<_Alignment>(__destination, __source, __total_size, __rank, __stride);
-        return async_contract_fulfillment::none;
-    }
-};
+// __try_get_barrier_handle returns barrier handle of block-scoped barriers and a nullptr otherwise.
+template<thread_scope _Sco, typename _CompF>
+_LIBCUDACXX_INLINE_VISIBILITY
+_CUDA_VSTD::uint64_t * __try_get_barrier_handle(barrier<_Sco, _CompF> & __barrier) {
+    return nullptr;
+}
 
 template<>
-struct __memcpy_async_impl<4, false> {
-    _LIBCUDACXX_DEVICE static inline async_contract_fulfillment __copy(char * __destination, char const * __source, _CUDA_VSTD::size_t __total_size, _CUDA_VSTD::size_t __rank, _CUDA_VSTD::size_t __stride) {
-        // Guard from host visibility when compiling in host only contexts
-        NV_IF_TARGET(
-            NV_IS_DEVICE, (
-                for (_CUDA_VSTD::size_t __offset = __rank * 4; __offset < __total_size; __offset += __stride * 4) {
-                    asm volatile ("cp.async.ca.shared.global [%0], [%1], 4, 4;"
-                        :: "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__destination + __offset))),
-                            "l"(__source + __offset)
-                        : "memory");
-                }
-            )
+_LIBCUDACXX_INLINE_VISIBILITY
+_CUDA_VSTD::uint64_t * __try_get_barrier_handle<::cuda::thread_scope_block, _CUDA_VSTD::__empty_completion>(barrier<::cuda::thread_scope_block> & __barrier) {
```

Comment:
I think this needs an `inline` to be a weak symbol; @miscco @wmaxey check what I'm saying here, but I think full function template specializations are treated as normal function definitions for the purposes of linkage?